plugins/LadspaEffect/LadspaEffect.cpp

```diff
@@ -129,26 +129,20 @@ void LadspaEffect::changeSampleRate()
 
 
 
-bool LadspaEffect::processAudioBuffer( SampleFrame* _buf,
-							const fpp_t _frames )
+double LadspaEffect::processImpl(SampleFrame* buf, const fpp_t frames)
 {
 	m_pluginMutex.lock();
-	if( !isOkay() || dontRun() || !isRunning() || !isEnabled() )
```

Comment:
You are un-complicating it, but I could imagine that it might have been that complicated on purpose?

Let's assume we enter the effect processing and stop right before this `if` (which is now inside `Effect.cpp`). Before  we reach the if, `LadspaEffect::changeSampleRate` kicks in, and when it re-instantiates the plugin, there are some errors, but before they can make the effect "not okay", this `!isOkay()` query (now in `Effect.cpp`) happens and thinks everything is OK. Now the instantiation finishes with a "not okay" effect and unlocks the `m_pluginMutex`, so we can pass the `m_pluginMutex.lock();` here. We then work with a "not okay" effect.

Is this a problem?

Reply:
Maybe the `Effect` class should have an optional plugin mutex that plugins can opt in to using during their processing loop? Or change the `processImpl` interface so it can tell the caller that effect processing did not occur?

Like this?
```
//! Returns true if processing occurred
bool processImpl(SampleFrame* buf, const fpp_t frames, double& outSum)
```

Reply:
I would avoid having a mutex in `Effect` since they might motivate some devs to use them, but they are not realtime safe. Not sure what you mean with the `Or...` part, can you please elaborate how this saves us from the described scenario?

Reply:
It would allow this in `processImpl`:
```
m_pluginMutex.lock();
if (!isOkay() || dontRun() || !isRunning() || !isEnabled())
{
    // Don't process
    return false;
}

// ... rest of method ...
```

If would check twice if processing should occur (once in `processAudioBuffer` and once in `processImpl`), but it would allow the mutex to protect the if-statement checking whether processing should occur.

I will look into this concurrency issue some more. The best case would be if the mutex is not actually needed, or not needed around that if-statement.

Reply:
Agreeing that this is a bit ugly, but I could live with it. Finding a replacement for a mutex sounds really hard. In Zyn, we would have solved this with something like preparing a pointer and then passing it to the realtime thread:

1. `LadspaEffect::changeSampleRate()` would create a separate LadspaEffect object, using `LadspaEffect::pluginInstantiation()` on it
2. It would simply pass the pointer to this object to the realtime thread - in Zyn, via OSC, but a simple pointer which is used like a mailbox could do it here.
3. Once the realtime thread again wants to run, before processing any audio, it looks if the pointer is set (i.e. if there is mail in the mailbox) - if yes, it does a pointer swap (this is realtime safe). In this case, we can't swap the whole `this` in one run, but we could probably quickly swap the members.
4. The realtime thread passes back the old object for destruction via another mailbox.
5. The non-realtime thread somehow has to care for calling `LadspaEffect::pluginDestruction`. This could be done lazily, e.g. with a slow `QTimer`.

I'm not sure if this really works here though, and if this is still in scope of this PR.

Reply:
This is fixed now